File: module-11-rl-agents/solutions/curriculum.ts
/**
 * Module 11: Curriculum learning
 * Start easy, promote to harder tasks as agent improves.
 */
import type { EvalTask } from "../../module-7-evals/solutions/types.js";

export interface CurriculumTier {
  name: string;
  difficulty: "easy" | "medium" | "hard";
  tasks: EvalTask[];
  promotionThreshold: number; // Pass rate to advance
}

export interface CurriculumState {
  currentTier: number;
  tierPassRates: number[];
  episodesCompleted: number;
  totalSuccesses: number;
}

export class Curriculum {
  private tiers: CurriculumTier[];
  private state: CurriculumState;

  constructor(tiers: CurriculumTier[]) {
    this.tiers = tiers;
    this.state = {
      currentTier: 0,
      tierPassRates: tiers.map(() => 0),
      episodesCompleted: 0,
      totalSuccesses: 0,
    };
  }

  getCurrentTier(): CurriculumTier {
    return this.tiers[this.state.currentTier];
  }

  /** Record an episode outcome and check for promotion */
  recordOutcome(success: boolean): { promoted: boolean; newTier?: string } {
    this.state.episodesCompleted++;
    if (success) this.state.totalSuccesses++;

    // Update running pass rate (exponential moving average)
    const alpha = 0.1;
    const current = this.state.tierPassRates[this.state.currentTier];
    this.state.tierPassRates[this.state.currentTier] =
      current * (1 - alpha) + (success ? 1 : 0) * alpha;

    // Check promotion
    const tier = this.tiers[this.state.currentTier];
    const passRate = this.state.tierPassRates[this.state.currentTier];

    if (passRate >= tier.promotionThreshold && this.state.currentTier < this.tiers.length - 1) {
      this.state.currentTier++;
      const newTier = this.tiers[this.state.currentTier];
      return { promoted: true, newTier: newTier.name };
    }

    return { promoted: false };
  }

  /** Pick a random task from the current tier */
  sampleTask(): EvalTask {
    const tier = this.getCurrentTier();
    return tier.tasks[Math.floor(Math.random() * tier.tasks.length)];
  }

  getState(): CurriculumState {
    return { ...this.state };
  }

  isComplete(): boolean {
    return (
      this.state.currentTier === this.tiers.length - 1 &&
      this.state.tierPassRates[this.state.currentTier] >=
        this.tiers[this.state.currentTier].promotionThreshold
    );
  }
}

File: module-11-rl-agents/solutions/python/curriculum.py
"""Module 11: Curriculum learning."""
import random


class Curriculum:
    def __init__(self, tiers: list[dict]):
        self.tiers = tiers
        self.state = {
            "current_tier": 0,
            "tier_pass_rates": [0.0] * len(tiers),
            "episodes_completed": 0,
            "total_successes": 0,
        }

    def get_current_tier(self) -> dict:
        return self.tiers[self.state["current_tier"]]

    def record_outcome(self, success: bool) -> dict:
        self.state["episodes_completed"] += 1
        if success:
            self.state["total_successes"] += 1

        alpha = 0.1
        idx = self.state["current_tier"]
        current = self.state["tier_pass_rates"][idx]
        self.state["tier_pass_rates"][idx] = current * (1 - alpha) + (1.0 if success else 0.0) * alpha

        tier = self.tiers[idx]
        pass_rate = self.state["tier_pass_rates"][idx]

        if pass_rate >= tier["promotion_threshold"] and idx < len(self.tiers) - 1:
            self.state["current_tier"] += 1
            new_tier = self.tiers[self.state["current_tier"]]
            return {"promoted": True, "new_tier": new_tier["name"]}

        return {"promoted": False}

    def sample_task(self):
        tier = self.get_current_tier()
        return random.choice(tier["tasks"])

    def get_state(self) -> dict:
        return dict(self.state)

    def is_complete(self) -> bool:
        idx = self.state["current_tier"]
        return (
            idx == len(self.tiers) - 1
            and self.state["tier_pass_rates"][idx] >= self.tiers[idx]["promotion_threshold"]
        )
